servicetalk-encoding-netty/src/main/java/io/servicetalk/encoding/netty/NettyChannelContentCodec.java

```diff
@@ -0,0 +1,342 @@
+                        if (next == END_OF_STREAM) {
+                            // May produce footer
+                            preparePendingData(channel);
+                            Buffer buffer = drainChannelQueueToSingleBuffer(channel.outboundMessages(), allocator);
+                            if (buffer != null) {
+                                subscriber.onNext(buffer);
+                            } else {
+                                subscriber.onNext(EMPTY_BUFFER);
+                            }
+
+                            return;
+                        }
+
+                        channel.writeOutbound(toByteBuf(next));
+                        Buffer buffer = drainChannelQueueToSingleBuffer(channel.outboundMessages(), allocator);
+                        if (buffer != null) {
+                            subscriber.onNext(buffer);
+                        } else {
+                            subscription.request(1);
+                        }
+                    }
+
+                    @Override
+                    public void onError(Throwable t) {
+                        safeCleanup(channel);
+                        subscriber.onError(t);
+                    }
+
+                    @Override
+                    public void onComplete() {
+                        try {
+                            cleanup(channel);
+                        } catch (Throwable t) {
+                            subscriber.onError(t);
+                            return;
+                        }
+
+                        subscriber.onComplete();
+                    }
+                });
+    }
+
+    @Override
+    public Buffer decode(final Buffer src, final int offset, final int length, final BufferAllocator allocator) {
+        requireNonNull(src);
+        final int availableBytes = src.readableBytes() - offset;
+        if (offset < 0) {
+            throw new IllegalArgumentException("Invalid offset: " + offset + " (expected >= 0)");
+        }
+
+        if (length > availableBytes) {
+            throw new IllegalArgumentException("Invalid length: " + length + " (expected <= " + availableBytes);
+        }
+
+        final ByteToMessageDecoder decoder = decoderSupplier.get();
+        final EmbeddedChannel channel = newEmbeddedChannel(decoder, allocator);
+
+        try {
+            ByteBuf origin = toByteBuf(src, offset, length);
+            channel.writeInbound(origin);
+
+            Buffer buffer = drainChannelQueueToSingleBuffer(channel.inboundMessages(), allocator);
+            if (buffer == null) {
+                throw new CodecDecodingException("Decoder didn't return any output.");
+            }
+            return buffer;
+        } catch (Throwable e) {
+            LOGGER.error("Error while decoding with {}", name(), e);
+            throwException(e);
+            return EMPTY_BUFFER;
+        } finally {
+            safeCleanup(channel);
```

Comment:
I believe you mean in case of successful completion then exception in safe-cleanup will hide it.
I added the cleanup in the try body, and left the safe one in finally. Any exception (including ones from cleanup) will surface unless something throws before cleanup, in which case that stays silent.

Reply:
If the `try` throws before `cleanup`, the catch block should do `safeCleanup` and the `finally` is not necessary. Trying to have a consistent approach with `encode` impl.

Reply:
Ignore this comment, saw the full picture after review: https://github.com/apple/servicetalk/pull/1376#discussion_r592547778